sdk/go/pulumi-language-go/main.go

```diff
@@ -331,31 +332,54 @@ func (host *goLanguageHost) GetRequiredPlugins(ctx context.Context,
 	}, nil
 }
 
-func execProgramCmd(cmd *exec.Cmd, env []string) error {
+func runCmdStatus(cmd *exec.Cmd, env []string) (int, error) {
 	cmd.Env = env
 	cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
 
 	err := cmd.Run()
-	if err == nil {
-		// Go program completed successfully
-		return nil
-	}
 
 	// error handling
 	exiterr, ok := err.(*exec.ExitError)
 	if !ok {
-		return errors.Wrapf(err, "command errored unexpectedly")
+		return 0, errors.Wrapf(err, "command errored unexpectedly")
 	}
 
 	// retrieve the status code
 	status, ok := exiterr.Sys().(syscall.WaitStatus)
 	if !ok {
-		return errors.Wrapf(err, "program exited unexpectedly")
+		return 0, errors.Wrapf(err, "program exited unexpectedly")
+	}
+
+	return status.ExitStatus(), nil
+}
+
+func runProgram(bin string, env []string) *pulumirpc.RunResponse {
+	cmd := exec.Command(bin)
+	status, err := runCmdStatus(cmd, env)
+	if err != nil {
+		return &pulumirpc.RunResponse{
+			Error: err.Error(),
+		}
+	}
+
+	if status == 0 {
+		return &pulumirpc.RunResponse{}
 	}
 
-	// If the program ran, but exited with a non-zero error code. This will happen often, since user
-	// errors will trigger this.  So, the error message should look as nice as possible.
-	return errors.Errorf("program exited with non-zero exit code: %d", status.ExitStatus())
+	// If the program ran, but returned an error,
+	// the error message should look as nice as possible.
+	if status == constant.ExitStatusLoggedError {
+		// program failed but sent an error to the engine
+		return &pulumirpc.RunResponse{
+			Bail: true,
+		}
+	}
+
+	// If the program ran, but exited with a non-zero and non-reserved error code.
+	// indicate to the user which exit code the program returned.
+	return &pulumirpc.RunResponse{
+		Error: fmt.Sprintf("program exited with non-zero exit code: %d", status),
```

Comment:
Stylistically (nit) it would be better for happy case to go to the bottom of the func, but this reads correct!